lattices/src/algebra.rs

```diff
@@ -57,14 +57,45 @@ pub fn ring<S: Debug + PartialEq + Clone, const N: usize>(
     inverse(items, f, zero, b);
 }
 
+/// Defines an integral domain structure.
+/// An integral domain is a nonzero commutative ring with no nonzero zero divisors.
+pub fn integral_domain<S: Debug + PartialEq + Clone, const N: usize>(
+    items: &[S; N],
+    f: &impl Fn(S, S) -> S,
+    g: &impl Fn(S, S) -> S,
+    zero: S,             // zero is the identity element of f
+    one: S,              // one is the identity element of g
+    b: &impl Fn(S) -> S, /* b is the function to compute the inverse element of an element with respect to f */
+) {
+    commutative_ring(items, f, g, zero.clone(), one, b);
+    no_nonzero_zero_divisors(items, g, zero);
+}
+
+/// Defines a no-nonzero-zero-divisors property.
+/// x is a nonzero divisor if xy = 0 and y is also a nonzero element.
+pub fn no_nonzero_zero_divisors<S: Debug + PartialEq + Clone, const N: usize>(
+    items: &[S; N],
+    f: &impl Fn(S, S) -> S,
+    zero: S,
+) {
+    for a in items {
+        for b in items {
+            if *a != zero && *b != zero {
+                assert_ne!(f(a.clone(), b.clone()), zero);
```

Comment:
Instead of panicking, we should return a `Result<(), (S, S)>` here, returning nothing if it passes but the counterexample if it fails. `panic` isn't quite the equivalent of an exception in the other language, it is meant for situations where there is an unrecoverable error and the program needs to halt. Instead, `Result` lets us have errors that can be handled by a parent.

Reply:
I have a PR out for the switch to Results as the return type. I will check it in after this PR is checked in since it is going to introduce a ton of merge conflicts